src/charm.py

```diff
@@ -1,117 +1,282 @@
+        if not self._are_relations_ready():
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        hostname = self.get_hostname()
+        if not hostname:
+            self.unit.status = BlockedStatus("Waiting for hostname to be configured")
+            event.defer()
+            return
+        self._tls.certificate_relation_joined(hostname, self.certificates)
+
+    def _on_certificates_relation_broken(self, _: Any) -> None:
+        """Handle the TLS Certificate relation broken event."""
+        self._reconcile()
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handle the TLS Certificate available event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        self._tls.certificate_relation_available(event)
+        self._reconcile()
+
+    def _on_certificate_expiring(
+        self,
+        event: Union[CertificateExpiringEvent, CertificateInvalidatedEvent],
+    ) -> None:
+        """Handle the TLS Certificate expiring event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        if not self._are_relations_ready():
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        self._tls.certificate_expiring(event, self.certificates)
+
+    # This method is too complex but hard to simplify.
+    def _certificate_revoked(self, revoke_list: List[str]) -> None:  # noqa: C901
+        """Handle TLS Certificate revocation.
+
+        Args:
+            revoke_list: TLS Certificates list to revoke
+        """
+        if not self._are_relations_ready():
+            return
+        tls_certificates_relation = self._tls.get_tls_relation()
+        for hostname in revoke_list:
+            old_csr = self._tls.get_relation_data_field(
+                f"csr-{hostname}",
+                tls_certificates_relation,  # type: ignore[arg-type]
+            )
+            if not old_csr:
+                continue
+            if JujuVersion.from_environ().has_secrets:
+                try:
+                    secret = self.model.get_secret(label=f"private-key-{hostname}")
+                    secret.remove_all_revisions()
+                except SecretNotFoundError:
+                    LOGGER.warning("Juju secret for %s already does not exist", hostname)
+                    continue
+            try:
+                self._tls.pop_relation_data_fields(
+                    [f"key-{hostname}", f"password-{hostname}"],
+                    tls_certificates_relation,  # type: ignore[arg-type]
```

Comment:
Nitpick: would type cast be helpful here?

Reply:
I'm not sure I see where could a type cast can be used, this method simply removes the 2 keys from the relation databag. Can you elaborate a bit on this?

Reply:
I'm closing this conversation for now, we can continue our discussion internally if needed